text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+* `initializers` - INITIALIZER
+* `instance-initializers` - INSTANCE-INITIALIZER
+* `models` - MODEL, ADAPTER, SERIALIZER
+* `partials` - PARTIAL
+* `routes` - ROUTE, CONTROLLER, template
+* `services` - SERVICE
+* `transforms` - TRANSFORM
+* `utils` - UTIL
+
+> Note: ALL CAPS indicates which collections are definitive (rule 3) for a type.
+
+The following default types are proposed for collections (rule 4):
+
+* `components` - component
+* `initializer` - initializer
+* `instance-initializers` - instance-initializer
+* `models` - model
+* `partials` - partial
+* `routes` - route
+* `services` - service
+* `transforms` - transform
+* `utils` - util
+
+The following private collections are allowed within collections (rule 5):
+
+* `components` - utils
+* `models` - utils
+* `initializers` - utils
+* `instance-initializers` - utils
+* `routes` - components, utils
+* `services` - utils
+* `transforms` - utils
+
+The following groups are proposed for collections (rule 6):
+
+* `data` - models, transforms
+* `init` - initializers, instance-initializers
+* `ui` - components, partials, routes
+
+The collection and type system is designed to be extensible, so that addons can
+contribute their own collections and types. The `data` collection and its
+corresponding types should be defined in ember-data. Liquid-fire might want to
+define an `animations` collection and a `transition` type, and expand `routes`
+to allow `animations` as a private collection.
+
+The specific format of collection and type declarations for addons is TBD.
+
+#### "Components" Collection
+
+This proposal broadens the scope of the term "component" to include all
+template-invocable parts of Ember. This includes today's components and helpers,
+and the future implementation of "glimmer components" (with angle brackets) and
+element modifiers.
+
+Grouping template-invocable elements together in a single collection recognizes
+that they already coexist in the same namespace. After all, only one helper OR
+component can be invoked as `{{foo-bar}}`. Using a common collection will not
+only simplify file management and searching, it will also provide implicit
+linting against creating a helper and class-based component of the same name.
+
+#### Private Collections
+
+You may wish to make a component available in a particular template without
+polluting the top-level `components` collection with a more local concern.
+Private collections allow you to augment a top-level collection's contents for
+use at a particular namespace.
+
+Private collections are declared as a directory sharing the name of the
+top-level collection, prefixed with a `-`. So the top-level `routes`
+collection could be augmented via a private `-components` collection.
+
+Say that you want to define a `post-viewer` component to be available only from
+within `src/ui/routes/posts/post/template.hbs`. You could achieve this by
+creating your component module in
+`src/ui/routes/posts/post/-components/post-viewer.js`.
+
```

Comment:
> Let's say I have {{#foo}}{{bar}}{{/foo}} in the /baz route template, to locate the {{bar}} component I'd expect it to be in src/ui/components/bar or perhaps src/ui/components/foo/bar (local lookup).

As a concept, local lookup is always talking about the invoking template. So in your example, you would not be looking in `src/ui/components/foo/bar` for this (since `{{bar` is invoked from the baz route template not from the foo component template).

This RFC is proposing the concept of local lookup (which is named private collections). This will require looking in exactly two locations for a given template invocation, global (`src/ui/components`) and local (`<current-template-location>/-components`). 

Ultimately, I believe that we will transition to using private collection based components by default and only make global components when something truly is global. As you mentioned, using private collections / local lookup is completely up to you, but I believe the benefits far outweigh the negatives.